### PR TITLE
add sponge construction

### PR DESCRIPTION
**Pull request by bkomuves**

### CODE REVIEW

File: poseidon2.nim [hunk 1]
```diff
@@ -56,13 +16,64 @@ func permInplace*(x, y, z : var F) =
...
+  let b = high(xs)
+  let n = b-a+1
+
+  var s0 : F = zero
+  var s1 : F = zero
+  var s2 : F = zero
+
+  for i in 0..<n:
+    s0 += xs[a+i]
+    permInplace(s0,s1,s2)
```
Comment (markspanbroek) [inline poseidon2.nim]:
```suggestion
  for x in xs:
    s0 += x
    permInplace(s0,s1,s2)
```

File: tests/poseidon2/testPoseidon2.nim [hunk 2]
```diff
@@ -23,6 +53,24 @@ suite "poseidon2":
     check toDecimal(y) == "09030699330013392132529464674294378792132780497765201297316864012141442630280"
     check toDecimal(z) == "09137931384593657624554037900714196568304064431583163402259937475584578975855"
 
+  test "sponge with rate=1":
+    for n in 0..7:
+      var xs: seq[F]
+      for i in 1..n:
+        xs.add( toF(i) )
+      let h = spongeWithRate1(xs)
+      # echo(toDecimal(h))
```
Comment (markspanbroek) [inline tests/poseidon2/testPoseidon2.nim]:
```suggestion
```

File: tests/poseidon2/testPoseidon2.nim [hunk 3]
```diff
@@ -23,6 +53,24 @@ suite "poseidon2":
...
+      check toDecimal(h) == expectedSpongeResultsRate1[n] 
+
+  test "sponge with rate=2":
+    for n in 0..7:
+      var xs: seq[F]
+      for i in 1..n:
+        xs.add( toF(i) )
+      let h = spongeWithRate2(xs)
+      # echo(toDecimal(h))
```
Comment (markspanbroek) [inline tests/poseidon2/testPoseidon2.nim]:
```suggestion
```

File: poseidon2/types.nim [hunk 4]
```diff
@@ -18,9 +18,17 @@ func getZero*() : F =
   setZero(z)
   return z
 
+func getOne*() : F = 
+  var y : F
+  # y.fromUint(1'u32)       # WTF, why does this not compile ???
```
Comment (markspanbroek) [inline poseidon2/types.nim]: The implementation of `fromUint` uses a cast from uint32 to array of byte, which isn't supported at compile time. The statement `const one*  : F = getOne()` in `roundfun.nim` is evaluated at compile time.

File: poseidon2/types.nim [hunk 5]
```diff
@@ -18,9 +18,17 @@ func getZero*() : F =
...
+  y.fromHex("0x01")
+  return y
+
+# for some reason this one does not compile... ???
+# (when actually called)
```
Comment (markspanbroek) [inline poseidon2/types.nim]: It is called in `testPoseidon2.nim` ?

**Review by markspanbroek — APPROVED** [review]
I can't really comment on the sponge construction itself, but I left a few comments regarding nim.

File: poseidon2/types.nim (hunk 4 above)
Comment (bkomuves) [reply poseidon2/types.nim]: Ah... And why is this simple casting not supported at compile time?

File: poseidon2/types.nim (hunk 5 above)
Comment (bkomuves) [reply poseidon2/types.nim]: Yeah this was the same issue as above (it's not working at compile time)

File: poseidon2/types.nim (hunk 4 above)
Comment (markspanbroek) [reply poseidon2/types.nim]: I don't know

File: poseidon2.nim [hunk 6]
```diff
@@ -56,13 +16,64 @@ func permInplace*(x, y, z : var F) =
...
+  var s0 : F = zero
+  var s1 : F = zero
+  var s2 : F = toF(0x0301) ; s2 += twoToThe64        # domain separation IV := (2^64 + 256*t + r)
+
+  for x in xs:
+    s0 += x
+    permInplace(s0,s1,s2)
+
+  # padding
+  s0 += one;
```
Comment (dryajov) [inline poseidon2.nim]:
```suggestion
  s0 += one
```

File: poseidon2/roundfun.nim [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+#-------------------------------------------------------------------------------
+
+const externalRoundConst : array[24, F] = arrayFromHex( externalRoundConstStr )
+const internalRoundConst : array[56, F] = arrayFromHex( internalRoundConstStr )
+
+#-------------------------------------------------------------------------------
+
+# inplace sbox, x => x^5
+func sbox*(x: var F) : void =
```
Comment (dryajov) [inline poseidon2/roundfun.nim:16]: This could be `templates`. Same for the rest of the `funcs` in this file.

File: poseidon2.nim [hunk 8]
```diff
@@ -56,13 +16,64 @@ func permInplace*(x, y, z : var F) =
...
   permInplace(x, y, z)
   return (x,y,z)
 
 #-------------------------------------------------------------------------------
 
+# sponge with rate=1 (capacity=2) 
+func spongeWithRate1*(xs: openArray[F]) : F =
+  let a = low(xs)
+  let b = high(xs)
+  let n = b-a+1
```
Comment (dryajov) [inline poseidon2.nim]: not used?

File: poseidon2.nim [hunk 9]
```diff
@@ -56,13 +16,64 @@ func permInplace*(x, y, z : var F) =
...
+# the Poseidon2 permutation
 func perm*(xyz: S) : S =
   var (x,y,z) = xyz
   permInplace(x, y, z)
   return (x,y,z)
 
 #-------------------------------------------------------------------------------
 
+# sponge with rate=1 (capacity=2) 
+func spongeWithRate1*(xs: openArray[F]) : F =
```
Comment (dryajov) [inline poseidon2.nim:30]: Will `xs` be fixed or arbitrary size?

Comment (bkomuves) [reply poseidon2.nim:30]: arbitrary, this is a "traditional" hash function.

File: poseidon2.nim (hunk 8 above)
Comment (bkomuves) [reply poseidon2.nim]: Mark suggested a refactoring, and I forgot to delete these. It was used before (the same way as in `spongeRateWith2`)

File: poseidon2/roundfun.nim (hunk 7 above)
Comment (elcritch) [reply poseidon2/roundfun.nim:16]: I did some earlier profiling with this and found that changing some of these funcs into templates could actually slow things down. Using `func` and letting the C compiler decide when to inline makes sense without a bunch of profiling.

Comment (dryajov) [reply poseidon2/roundfun.nim:16]: You're right, the compiler might be smarter about this.

Comment (elcritch) [reply poseidon2/roundfun.nim:16]: Yah seems the machines won't leave us any areas where they're not smarter. ;)

**Review by dryajov — CHANGES_REQUESTED** [review]
LGTM, with the suggested changes.

**Comment by markspanbroek** [comment]
I've addressed @dryajov's comments and merged this, I hope you don't mind @bkomuves.

**Pull request merged**